File: ui/src/sse.ts
// Reader for the server's SSE frames: "id: N\nevent: T\ndata: J\n\n"
// (internal/server/server.go writeSSEFrame). fetch + a hand-written parser
// instead of EventSource, which cannot send the Authorization header and
// reconnects forever once a finished turn's stream closes.

export interface Frame {
  id: number | null
  event: string
  data: string
}

// parseFrames splits buf into complete frames and returns the unparsed
// remainder, so a chunk boundary in the middle of a frame is safe.
export function parseFrames(buf: string): { frames: Frame[]; rest: string } {
  const frames: Frame[] = []
  let start = 0
  for (;;) {
    const end = buf.indexOf('\n\n', start)
    if (end < 0) break
    const frame = parseFrame(buf.slice(start, end))
    if (frame) frames.push(frame)
    start = end + 2
  }
  return { frames, rest: buf.slice(start) }
}

function parseFrame(block: string): Frame | null {
  const frame: Frame = { id: null, event: '', data: '' }
  const data: string[] = []
  for (const line of block.split('\n')) {
    if (line.startsWith('id: ')) frame.id = Number(line.slice(4))
    else if (line.startsWith('event: ')) frame.event = line.slice(7)
    else if (line.startsWith('data: ')) data.push(line.slice(6))
  }
  frame.data = data.join('\n')
  return frame.event ? frame : null
}

export async function* readSSE(body: ReadableStream<Uint8Array>): AsyncGenerator<Frame> {
  const reader = body.getReader()
  const decoder = new TextDecoder()
  let buf = ''
  try {
    for (;;) {
      const { value, done } = await reader.read()
      if (done) return
      buf += decoder.decode(value, { stream: true })
      const parsed = parseFrames(buf)
      buf = parsed.rest
      for (const frame of parsed.frames) yield frame
    }
  } finally {
    reader.releaseLock()
  }
}

File: crates/otto-core/src/wire/sse.rs
//! Server-Sent Events framing.
//!
//! One frame is `id: <seq>\nevent: <type>\ndata: <json>\n\n`, written by
//! `writeSSEFrame` in `internal/server/server.go` and read by `parseFrames`
//! in `ui/src/sse.ts`.

use serde::{Deserialize, Serialize};

/// One decoded frame. `id` is absent when the frame carried no `id:` line.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Frame {
    pub id: Option<i64>,
    pub event: String,
    pub data: String,
}

/// What [`parse_frames`] returns: the complete frames in `buf` and the
/// trailing bytes that are not yet a whole frame.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ParsedFrames {
    pub frames: Vec<Frame>,
    pub rest: String,
}

/// Writes one frame exactly as the Go server does.
pub fn format_frame(seq: i64, event: &str, data: &str) -> String {
    format!("id: {seq}\nevent: {event}\ndata: {data}\n\n")
}

/// Splits `buf` on the blank line between frames. A block with no `event:`
/// line is dropped, matching `parseFrame`'s `return frame.event ? frame : null`.
pub fn parse_frames(buf: &str) -> ParsedFrames {
    let mut frames = Vec::new();
    let mut start = 0usize;
    while let Some(offset) = buf[start..].find("\n\n") {
        let end = start + offset;
        if let Some(frame) = parse_frame(&buf[start..end]) {
            frames.push(frame);
        }
        start = end + 2;
    }
    ParsedFrames {
        frames,
        rest: buf[start..].to_string(),
    }
}

fn parse_frame(block: &str) -> Option<Frame> {
    let mut id = None;
    let mut event = String::new();
    let mut data: Vec<&str> = Vec::new();
    for line in block.split('\n') {
        if let Some(rest) = line.strip_prefix("id: ") {
            // `Number("x")` is NaN in the TypeScript reader; Rust has no NaN
            // for an integer field, so an unparsable id reads as absent.
            id = rest.trim().parse::<i64>().ok();
        } else if let Some(rest) = line.strip_prefix("event: ") {
            event = rest.to_string();
        } else if let Some(rest) = line.strip_prefix("data: ") {
            data.push(rest);
        }
    }
    if event.is_empty() {
        return None;
    }
    Some(Frame {
        id,
        event,
        data: data.join("\n"),
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn splits_complete_frames_and_keeps_a_partial_one() {
        let parsed =
            parse_frames("id: 0\nevent: text_delta\ndata: {\"a\":1}\n\nid: 1\nevent: agent_fin");
        assert_eq!(
            parsed.frames,
            vec![Frame {
                id: Some(0),
                event: "text_delta".into(),
                data: "{\"a\":1}".into(),
            }]
        );
        assert_eq!(parsed.rest, "id: 1\nevent: agent_fin");
    }

    #[test]
    fn drops_a_block_without_an_event_line() {
        let parsed = parse_frames("id: 7\ndata: {}\n\n");
        assert!(parsed.frames.is_empty());
        assert_eq!(parsed.rest, "");
    }

    #[test]
    fn joins_repeated_data_lines_with_newlines() {
        let parsed = parse_frames("event: e\ndata: one\ndata: two\n\n");
        assert_eq!(parsed.frames[0].data, "one\ntwo");
        assert_eq!(parsed.frames[0].id, None);
    }

    #[test]
    fn format_frame_matches_the_go_writer() {
        assert_eq!(
            format_frame(3, "text_delta", "{\"text\":\"hi\"}"),
            "id: 3\nevent: text_delta\ndata: {\"text\":\"hi\"}\n\n"
        );
    }
}
